### PR TITLE
Create UserDefinedExpression feature

### PR DESCRIPTION
**Pull request by momosetkn**
You can define your own Expression.
For example, users can implement count-distinct with multiple columns, etc.

### CODE REVIEW

File: doma-core/src/main/java/org/seasar/doma/jdbc/criteria/expression/UserDefinedExpression.java [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    } else if (klass == java.lang.Long.class || klass == long.class) {
+      return new LongPropertyType(null);
+    } else if (klass == java.lang.Short.class || klass == short.class) {
+      return new ShortPropertyType(null);
+    } else if (klass == java.lang.String.class) {
+      return new StringPropertyType(null);
+    } else {
+      throw new UnsupportedOperationException("Does not support for " + klass.getName() + " type.");
+    }
+  }
```
Comment (momosetkn) [inline doma-core/src/main/java/org/seasar/doma/jdbc/criteria/expression/UserDefinedExpression.java:88]: For doma-kotlin, it will be primitive type

**Comment by nakamura-to** [comment]
Thank you for the pull request.
I plan to review it within a few days.

**Review by nakamura-to — CHANGES_REQUESTED** [review]

File: doma-core/src/main/java/org/seasar/doma/jdbc/criteria/expression/UserDefinedExpression.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+  private final Consumer<Declaration> block;
+
+  public UserDefinedExpression(Class<PROPERTY> klass, Consumer<Declaration> block) {
+    this.klass = Objects.requireNonNull(klass);
+    this.type = getTypeByKlass();
+    this.block = Objects.requireNonNull(block);
+  }
+
+  public UserDefinedExpression(
+      Class<PROPERTY> klass, EntityPropertyType<?, ?> type, Consumer<Declaration> block) {
```
[the remark this answers is not in the document]
Comment (momosetkn) [reply doma-core/src/main/java/org/seasar/doma/jdbc/criteria/expression/UserDefinedExpression.java:30]: Sorry, How do I obtain a class object using type? or
obtain from PropertyMetamodel?

[the remark this answers is not in the document]
Comment (nakamura-to) [reply doma-core/src/main/java/org/seasar/doma/jdbc/criteria/expression/UserDefinedExpression.java:30]: I was mistaken. I meant to say that by accepting a parameter of type `PropertyMetamodel`, both `Class` and `EntityPropertyType` can be resolved.

File: doma-core/src/main/java/org/seasar/doma/jdbc/criteria/expression/UserDefinedExpression.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+    public UserDefinedExpressionContext(
+        List<DeclarationItem> declarationItems, List<Object> parameters) {
+      this.declarationItems = declarationItems;
+      this.parameters = parameters;
+    }
+  }
+
+  /** Declaration of UserDefinedExpression */
+  public static class Declaration implements SqlContext {
```
[the remark this answers is not in the document]
Comment (momosetkn) [reply doma-core/src/main/java/org/seasar/doma/jdbc/criteria/expression/UserDefinedExpression.java:117]: Is it okay to omit appendParameter?
I didn't have a use-case for it, and I can almost achieve the same functionality with Expressions.literal, plus I wasn't meeting the test coverage.

[the remark this answers is not in the document]
Comment (nakamura-to) [reply doma-core/src/main/java/org/seasar/doma/jdbc/criteria/expression/UserDefinedExpression.java:117]: > Is it okay to omit appendParameter?

Yes, it's okay.
If someone makes a request in the future, we will consider it.

**Comment by momosetkn** [comment]
@nakamura-to 
I'm sorry I'm late.
Refactoring is complete.
I implemented the equals and hashCode method in UserDefinedExpression, similar to the others.

**Comment by nakamura-to** [comment]
Thank you!

**Pull request merged**